src/Fixer/Operator/OperatorLinebreakFixer.php

```diff
@@ -170,17 +170,16 @@ private function getExcludedIndices(Tokens $tokens): array
     {
         $colonIndices = [];
 
+        /** @var SwitchAnalysis $analysis */
```

Comment:
I think `@var` annotations should be avoided as much as possible because currently PHPStan always assumes they are valid, which can lead to undetected errors: https://phpstan.org/r/fedebefa-00bc-4671-8762-8761ed013727.

Those annotations are almost always a symptom of a missing/incorrect type information somewhere else, maybe in the return type of `ControlCaseStructuresAnalyzer::findControlStructures()`?

Reply:
`findControlStructures()` returns `Generator<int, AbstractControlCaseStructuresAnalysis>`. As far as I understand, in this case it must return `SwitchAnalysis[]` because it receives `type: [T_SWITCH]` argument, so it's not possible to narrow return type without creating a PHPStan extension. I considered replacing it with `assert($analysis instanceof SwitchAnalysis)` (see doctrine/coding-standard#47), but `assert()` doesn't seem to be used in this project anywhere else.

Reply:
Maybe changing the return type to `iterable<int, AbstractControlCaseStructuresAnalysis>` would be more suitable?

Reply:
It would make the return type wider, which is the opposite of the point of this PR (although I don't think it would make any difference in this specific case). Are there any benefits of using `iterable` instead of `Generator` in docblocks?

Reply:
> We could dedice to return an `array` instead without changing the return type.

We call `iterator_to_array()` on the result in tests, so it would break if we return `array` instead of `Generator`.

If this class implemented an interface, it would be nice to have `iterable` return type in the interface and `Generator` in the implementation, but that's not the case.

> Or maybe the actual result is a list of `AbstractControlCaseStructuresAnalysis`, then `list<AbstractControlCaseStructuresAnalysis>` would be best.

Psalm and PHPStan treat `list<Foo>` like `array<int, Foo>`, so we can't use it, although it's easier to understand for humans.

> But the main point here is to avoid the need to use `@var` where the function is called, I don't know if iterable or list would solve that.

No, we would need a PHPStan extension just for this single method, so it's not worth it. What do you think about replacing `@var` with [`assert()`](https://www.php.net/manual/en/function.assert.php) call? This way the type would be verified at runtime.

Reply:
The issue is that PHPStan allows `mixed->getCases()` (method may be there), but not `AbstractControlCaseStructuresAnalysis->getCases()` (there's no such method in this class, only in one of it's children). After changing return type of `ControlCaseStructuresAnalyzer::findControlStructures()` from `mixed[]` to `AbstractControlCaseStructuresAnalysis[]` it throws error, so we need to tell it that in this case it's always `SwitchAnalysis[]`.

We don't need a runtime check now, but if the assumption behind the `@var` annotation becomes invalid in the future, PHPStan will blindly believe it anyway. That's why I think it would be better to use `assert()`, so that invalid assumptions could be discovered by tests.

Also, please note that `$analysis instanceof SwitchAnalysis` check was there before (line 178). I'm not sure why it's there (can `$analysis` be something else? why is it after `->getCases()` call?).

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/51c8146236ef4911a216b03d729d39d26cdf4a6a/src/Fixer/Operator/OperatorLinebreakFixer.php#L173-L185

Reply:
I think the instance check was there to silent the SCA in the first place. Again, I share the analysis of the issue, but I think using `@var` is better than `assert`. As there is no dynamic part here we can test through unit tests alone and do not need dynamic tests, like `assert` and also (agreed) the `instance` check.

>if the assumption behind the @var annotation becomes invalid in the future

I don't share that idea.

Still makes me wonder, why does this need changing, what is wrong with the current code + SCA setup here?

Reply:
> I think the instance check was there to silent the SCA in the first place.

That's good, I was afraid that it might be necessary, although I didn't see how it possibly could :)

> > if the assumption behind the `@var` annotation becomes invalid in the future
>
> I don't share that idea.

OK, I think we need to leave it like that then (or `if (!$analysis instanceof SwitchAnalysis) continue;`).

> Still makes me wonder, why does this need changing, what is wrong with the current code + SCA setup here?

It's the issue I described in the first paragraph of my previous comment. `findControlStructures()` return type got more specific. PHPStan allowed calling a method on `mixed` ([before](https://phpstan.org/r/c574aebe-9bca-4693-9f3e-f9abf21e4447)), but not on an abstract class that might not have this method ([after](https://phpstan.org/r/b8fa3a7c-cdab-4524-b43c-d541c76f5811)), so we need to specify the exact class, as, in this case, we know it ([current solution with `@var`](https://phpstan.org/r/b707fe13-8cc2-40d2-861e-b4ec26f71fc2)/[instance check](https://phpstan.org/r/5a08ec9f-29b1-4182-8953-98369287933d)).

Reply:
>  given that you pass a `T_SWITCH` token you will always get a `SwitchAnalysis`

I misunderstood this part and now I get why we would need a PHPStan extension 😅. I agree it's not worth it.

What about moving the `instanceof` check to the beginning of the `foreach` loop? It seems to be there only to make PHPStan happy but it's an acceptable trade-off IMO.

Reply:
If we have runtime check anyway, why not in `assert()`?

I spent a lot of time wondering why there was this `if` statement in the first place. I've read through `findControlStructures()` code, which is ~150 LoC, and it isn't obvious that it has to return `SwitchAnalysis[]` given you pass `T_SWITCH`. I've added `var_dump()` and ran tests. It was always `SwitchAnalysis[]`, but I still wasn't certain that it can't be something else until @SpacePossum made it clear.

`assert()`, on the other hand, clearly communicates that the condition should be always true and that assertion is there only to make SCA happy. If it was there in the first place, I wouldn't waste my time on all of that and just assume that the return type of `findControlStructures()` depends on `$types` argument.

Reply:
>If we have runtime check anyway,

It is just there to keep the SCA happy, I would much more prefer no runtime check at all, because we do not need it as there is no dynamic component, as such we can test it with unit tests.
For readability I prefer comments over runtime checks as well.

Reply:
Both `@var`  and `assert()` communicate that the condition is always true, whereas `if` communicates that it may be false, which it can't, that's why I don't want to leave it there. So, while `assert()` is safer IMO, I'm also OK with `@var`. @julienfalque, what do you think?

Reply:
I would still reject `@var` as it cannot really be trusted in the context of PHPStan for now.

I never used `assert()` but, as far as I understand, its behavior changes depending on some PHP settings. I'm afraid that PHP CS Fixer will mostly be used in a context where those settings will make `assert()` have unwanted side-effects (warning or exception). Maybe we can change those settings in PHP CS Fixer's bootstrap code? This feels a bit hacky though, IMO.

Another solution could be to just ignore this error in `phpstan.neon` with a comment that explains why. We could also drop the `instanceof` check.

Reply:
> as far as I understand, its behavior changes depending on some PHP settings

That's correct, depending on [`zend.assertions`](https://www.php.net/manual/en/ini.core.php#ini.zend.assertions) setting it is either executed and throws `AssertionError` or will be ignored. This way you can have assertions executed in dev environment/during tests, but turn them off for production.

> Maybe we can change those settings in PHP CS Fixer's bootstrap code?

There's no need for that. It's important that the assertions would be executed during tests, which would make them trustworthy, unlike `@var`. We don't care whether `assert()` is enabled by end users, as we already know that the condition is always correct.

> I'm afraid that PHP CS Fixer will mostly be used in a context where those settings will make `assert()` have unwanted side-effects (warning or exception).

That would mean that our assumption is broken so an exception should be thrown. That's the point of using `assert()` instead of `@var`.

> Another solution could be to just ignore this error in `phpstan.neon` with a comment that explains why.

With `@var` or `assert()` PHPStan can “recover” and e.g. analyze whether we pass correct arguments to `->getCases()` method. Ignoring errors like unknown method call would mean that it wouldn't be able to analyse anything in this part of code.

Reply:
> That would mean that our assumption is broken so an exception should be thrown.

Indeed, then I think `assert()` is the best solution.

Reply:
```
We don't care whether assert() is enabled by end users, as we already know that the condition is always correct.
```

for that reason, why not just test in a unit test like we already for this code and do with everything else we know can be statically tested? Why add a runtime check and/or test-code (which `assert` is) into the code?

If it is for readability, than we can use comments, no runtime checks needed, no config for `assert` behavior. 

I'm very much -1 adding code for test and/or SCA tooling to code, so if we cannot solve it with comments or PHPDoc's than lets add it to some baseline and move on.

Reply:
>our assumption is broken 

we do not assume, we _know_ through unit tests